Backport zoneinfo logic into dateutil.tz.tzfile

pganssle opened the pull request:
## Summary of changes

This moves `tzfile` into its own submodule containing what is effectively a backport of the pure python implementation of `zoneinfo.ZoneInfo`, but with the semantics of a `dateutil.tz.tzfile` object.

The major differences from zoneinfo are:

1. All caching logic is implemented in `tz.gettz` rather than in `tz.tzfile`.
2. `tzfile` objects are pickled by value, not by reference to the IANA key.
3. `tzfile` has equality-by-value semantics, whereas `zoneinfo` has equality-by-identity semantics.

This does change the internal implementation details of the class. It also fixes at least two bugs:

1. It adds support for V2 and V3 files, which also solves the 2038 problem (and the "slim zoneinfo" problem).
2. Prior to this change, `dateutil` did not use the `fold` attribute during gaps (only during folds), in violation of PEP 495. This fixes that, but that does change the semantics of imaginary times a bit (and thus may affect some mechanisms for detecting imaginary times).

This is mostly a code dump with compatibility adjustments. It's quite possible that we'll want to refactor, particularly with respect to the way POSIX strings are handled. Theoretically `tzfile` could fall back to a `tz.tzstr` or some other `tz.tzrange`, or we could refactor `tzstr` and `tzrange` in terms of _CalendarOffset, _DayOffset and _TZStr (or both, by refactoring the internal logic of `tzstr` and/or `tzrange`).

Closes #462, #1059

**Open Tasks**:
- [x] Add the `key` parameter for consistency with `zoneinfo.ZoneInfo`
- [x] Add support for customizing `TZPATH`
- [x] Backport relevant tests from zoneinfo
    - [X] Basic tests
    - [x] Weird zone tests
    - [x] TZStr-specific tests
    - [x] TZPath tests
    - [x] V1 support tests
- [x] Fix issue with the repr when falling back to `tzdata`
- [x] Add tests against the "slim" tzdata builds.
- [x] Format all new and changed code with `black` (at least trying to be incremental here when we can...)
- [x] Document all relevant changes

### Pull Request Checklist
- [x] Changes have tests
- [x] Authors have been added to [AUTHORS.md](https://github.com/dateutil/dateutil/blob/master/AUTHORS.md)
- [x] News fragment added in changelog.d. See [CONTRIBUTING.md](https://github.com/dateutil/dateutil/blob/master/CONTRIBUTING.md#changelog) for details

pganssle committed:
Add test for PEP 495 gap logic in tzfile

PEP 495 says that when faced with an ambiguous choice (in a fold or a
gap), a tzinfo should return the offset from *before* the transition
when fold is 0 and the offset from *after* the transition when fold is
1, which affects both gaps (imaginary times) and folds (ambiguous
times).

Currently, `dateutil.tz.tzfile` only properly handles ambiguous times,
not imaginary times.

This adds xfailing tests for the proper handling of transitions with a
gap.

pganssle committed:
Remove test of implementation details

This tests implementation details of `tzfile` which are subject to
change. The values read for `isstd` only really matter insofar as they
affect the results of `utcoffset()` calls, which should be well covered
by other tests. Given that we are going to change these implementation
details anyway, it's time to remove this test.

pganssle committed:
Backport zoneinfo logic into tzfile

This moves `tzfile` into its own submodule containing what is
effectively a backport of the pure python implementation of
`zoneinfo.ZoneInfo`, but with the semantics of a `dateutil.tz.tzfile`
object.

The major differences from zoneinfo are:

1. All caching logic is implemented in `tz.gettz` rather than in
   `tz.tzfile`.
2. `tzfile` objects are pickled by value, not by reference to the IANA
   key.
3. `tzfile` has equality-by-value semantics, whereas `zoneinfo` has
   equality-by-identity semantics.

This does change the internal implementation details of the class. It
also fixes at least two bugs:

1. It adds support for V2 and V3 files, which also solves the 2038
   problem (and the "slim zoneinfo" problem).
2. Prior to this change, `dateutil` did not use the `fold` attribute
   during gaps (only during folds), in violation of PEP 495. This fixes
   that, but that does change the semantics of imaginary times a bit
   (and thus may affect some mechanisms for detecting imaginary times).

This is mostly a code dump with compatibility adjustments. It's quite
possible that we'll want to refactor, particularly with respect to the
way POSIX strings are handled. Theoretically `tzfile` could fall back to
a `tz.tzstr` or some other `tz.tzrange`, or we could refactor `tzstr`
and `tzrange` in terms of _CalendarOffset, _DayOffset and _TZStr (or
both, by refactoring the internal logic of `tzstr` and/or `tzrange`).

mariocj89 commented:
@pganssle do you want me to review this?

pganssle commented:
@mariocj89 I still need to backport the tests, but you are welcome to review it. I expect in the long run we'll want to consolidate some of the logic around `tzstr` and `tzrange` and some of these custom rule representations built into the TZif parser (for the extended version 2 support that allows specifying a POSIX string), but I don't expect to overhaul anything else.

Also I believe some documentation changes will be necessary before we can merge this as well.

mariocj89 commented:
OK, I'll take a look once all is ready.

pganssle committed:
Add support for key property

pganssle committed:
Add backport for a bug in os.path.isfile

On Python <3.7, `os.path.isfile` returns `False` with non-UTF8 paths
rather than raising an error, see ttps://bugs.python.org/issue33721

Jorricks commented:
So @mariocj89 I suppose this is really the only remaining blocker for dropping support for Python2?
Given this has been open for ages, should I try to prioritize this MR, open a new MR and see if we can get somewhere within the next few months?

mariocj89 commented:
@Jorricks this is indeed blocking most of the progress in the repo. If you have cycles to work on this we could drop Python 2 after we land and release this. @pganssle has been wanting to work on this for a long time but was not able to commit time to it.

kloczek commented:
+1

Jorricks commented:
Hey @mariocj89,
I'm attempting to start but I am missing some context. Could you help me out answering these questions :)?
1. How is this effort blocking the Python2 removal/deprecation effort?
2. If I understand correctly, the idea of this MR is to back port a copy of [zoneinfo](https://docs.python.org/3/library/zoneinfo.html) (that is only available in Python3.9+) into this application which is compatible with the current `tzinfo` class right?
3. Is the idea to implement this feature Python3.7+ or should it be Python2 compatible? Mostly asking because I'd like to add type hints already :).

mariocj89 commented:
> How is this effort blocking the Python2 removal/deprecation effort?

Without this change users in 2.7 will not get tz upgrades. This splits the tzdata from the package, allowing users to stay in an old version of the package but still get tzdata upgrades.

> If I understand correctly, the idea of this MR is to back port a copy of [zoneinfo](https://docs.python.org/3/library/zoneinfo.html) (that is only available in Python3.9+) into this application which is compatible with the current tzinfo class right?

Correct

> Is the idea to implement this feature Python3.7+ or should it be Python2 compatible? Mostly asking because I'd like to add type hints already :).

Needs to work for py2, so no type hints :).

Jorricks commented:
> > How is this effort blocking the Python2 removal/deprecation effort?
> 
> Without this change users in 2.7 will not get tz upgrades. This splits the tzdata from the package, allowing users to stay in an old version of the package but still get tzdata upgrades.
> 
> > If I understand correctly, the idea of this MR is to back port a copy of [zoneinfo](https://docs.python.org/3/library/zoneinfo.html) (that is only available in Python3.9+) into this application which is compatible with the current tzinfo class right?
> 
> Correct
> 
> > Is the idea to implement this feature Python3.7+ or should it be Python2 compatible? Mostly asking because I'd like to add type hints already :).
> 
> Needs to work for py2, so no type hints :).

Now it completely makes sense! Thanks.

Jorricks commented:
@mariocj89 to improve visibility on the progress & smoothen the process a little. Can we open a new branch containing this work so far?
We can just keep using this topic to keep track of everything that remains to be done so far I suppose.

I was diving a bit further (as I want to make sure I understand the full context before making changes) and was wondering the following:
During the building of the `python-dateutil` wheel, we package the `dateutil-zoneinfo.tar.gz`. This could(/should?) then also be replaced with a slim version of tzdata right?

mariocj89 commented:
@pganssle do you have some time to guide @Jorricks through this change please?

sfc-gh-izinkovsky commented:
@pganssle are you still actively working on getting this change merged?

Jorricks commented:
Unfortunately I don't have the time anymore to work on this and made zero progress anyway. So I don't think anyone is working on this. 

The code is very complex and you are only doing this for python 2 users. Given that reached end of life years ago, I'm not to enthusiastic to work on this anyway.

pganssle committed:
Add support for tzdata

pganssle committed:
Add available_iana_zones function

pganssle committed:
Add support for PYTHONTZPATH

pganssle committed:
TEMP: Fast-fail

pganssle committed:
Backport context-decorator functionality

pganssle committed:
Drop updatezinfo

pganssle committed:
Remove the updatezinfo script

pganssle committed:
Explicitly deprecate zoneinfo, remove from __all__

Removing `dateutil.zoneinfo` from `__all__` is a breaking change, but
I suspect that this is better than raising DeprecationWarning for
everyone who does `from dateutil import *`.

kloczek commented:
Is it anything else on pending list before  release new version? 🤔

pganssle committed:
Add test for TZPATH changes

This creates a fictional time zone Fictional/Liliput (a copy of
Arctic/Longyearbyne) and puts it in a temporary directory that is used
as the TZPATH, then we ensure that we can load the result.

pganssle commented:
@kloczek This is the main feature blocking 3.0.

My plan is:

1. Set up release automation
2. Make a 2.8.3 release with what is on `master` now
3. Make a 3.0.0 release that includes this change.
4. Remove support for Python <3.7-ish?
5. Make a 4.0.0 release.

pganssle commented:
Current status of this branch:

I believe all features are complete, we just need to backport some of the tests from `zoneinfo` and add a test that uses "slim" builds for `tzdata`, to ensure the TZif Version 2+ support is working.

If anyone has time to review the code now, no need to wait for the backported tests.

nanonyme commented:
Are those 3.5 tests failing just because legacy versions of Python 3 do not stringify (through os.fspath) pathlib objects so you have to do it yourself?

pganssle committed:
Fix available_iana_timezones when tzdata is missing

pganssle committed:
WIP: Start backporting tests

pganssle committed:
Skip some tests on rearguard

pganssle committed:
Use double quotes in tests/conftest.py

pganssle committed:
Fixes related to sub-minute offsets

pganssle committed:
Test fixes related to sub-minute offsets

pganssle committed:
Tests for various gaps and folds

pganssle committed:
Add tests for setting fold when converting from UTC

pganssle committed:
Add tests for datetime.time objects

pganssle committed:
Add test zone construction function

pganssle committed:
Add a test for a zone with 1 transition

pganssle committed:
Add a test with one DST -> DST transition

pganssle committed:
Add a test for no TZSTR set

pganssle committed:
Add test for zones with no transitions

pganssle committed:
Add test that empty zones raise

pganssle committed:
Add tests for very large transition values

pganssle committed:
Add tests for fixed offset zones with phantom transitions

pganssle committed:
Refactor weirdzone tests

pganssle committed:
Add tzstr tests

pganssle committed:
Add key to repr

pganssle commented:
Made a lot of progress on this at the sprints, but I don't think I'm going to quite get there today. I believe the main things left to do are:

1. Documentation of the changes
2. Test the tzfile with V1 data
3. Maybe adjust the repr to treat key/filename a bit better
4. Test the situation where only slim tzdata is installed

pganssle committed:
Add tzpath fixture

kytta commented:
Hey @pganssle, [you said](https://github.com/dateutil/dateutil/issues/653#issuecomment-1396351075) that this PR "is really not that hard for a motivated individual to finish". After two years since my last comment, I think I've become motivated enough :D

I've started by rebasing this branch against `master` to get rid of conflicts: https://github.com/kytta/dateutil/compare/new_tzfile. You can reset your branch to mine, if you wish.

However, I am kind of stuck on the list of "things left to do", so maybe you can comment on some things:

> 2. Test the tzfile with V1 data

I assume this means that V2 (and V3) tests are ready? If so, where can I find them? I am a bit lost looking at huge `test_tz.py` and `test_tzfile.py`... 😅 

Also, does this still mean that it would be enough to backport https://github.com/pganssle/zoneinfo/blob/666d80c27bda69541130758bbc7f1c9e035f79a0/tests/test_zoneinfo.py?

> 3. Maybe adjust the repr to treat key/filename a bit better

Do you mean `tzfile.__repr__` or something else?

> 4. Test the situation where only slim tzdata is installed

I guess I know how to do this in GitHub Actions, but do you have guidance on how to enable developers to test this, too?

---

I will take my time to properly understand the changes in this PR, so no rush with your reply :)

pganssle committed:
Improve testing for new tzfile implementation

This essentially backports the tests from the `zoneinfo` test suite.

pganssle committed:
Documentation for the new tzfile

pganssle committed:
Add changelog entry for #1130

pganssle committed:
Put tzwin as lower priority than tzfile

StanFromIreland commented on src/dateutil/tz/_tzfile.py:
```diff
@@ -0,0 +1,1093 @@
+            if not isdsts[comp_idx]:
+                dstoff = utcoff - utcoffsets[comp_idx]
+
+            if not dstoff and idx < (typecnt - 1):
```
See https://github.com/python/cpython/pull/145885/changes#diff-7c5f7d13590c56bce99efe1cd8d9692be0953d9adf1aee20b24e4192e7d42280R341

StanFromIreland commented on src/dateutil/tz/tz.py:
```diff
@@ -1625,35 +1082,38 @@ def nocache(name=None):
                             pass
                     else:
                         tz = None
+                        if not tz:
```
But you just set `tz = None`, so this is always true?

StanFromIreland commented on src/dateutil/__init__.py:
```diff
@@ -6,13 +6,13 @@
+    if name in __all__ or name == "zoneinfo":
         return importlib.import_module("." + name, __name__)
     raise AttributeError(
         "module {!r} has not attribute {!r}".format(__name__, name)
```
FYI, spotted an existing typo.

StanFromIreland commented on src/dateutil/__init__.py:
```diff
@@ -21,4 +21,6 @@ def __getattr__(name):
     # __dir__ should include all the lazy-importable modules as well.
-    return [x for x in globals() if x not in sys.modules] + __all__
+    return (
+        [x for x in globals() if x not in sys.modules] + __all__ + ["zoneinfo,"]
```
```suggestion
        [x for x in globals() if x not in sys.modules] + __all__ + ["zoneinfo"]
```
Did you mean?

StanFromIreland commented on src/dateutil/_tzdata_impl.py:
```diff
@@ -0,0 +1,89 @@
+else:
+    import importlib.resources
+
+    if sys.version_info > (3, 9):
```
```suggestion
    if sys.version_info >= (3, 9):
```
[It was added in 3.9, no?](https://docs.python.org/dev/library/importlib.resources.html#importlib.resources.files)

StanFromIreland commented on src/dateutil/_tzdata_impl.py:
```diff
@@ -0,0 +1,89 @@
+    """Indicates that a time zone file isn't present in tzdata."""
+
+    def __init__(self, msg, *args, **kwargs):
+        # TODO: Use keyword-only argument when this is Python 3-only
```
Python 2 usage has dropped severely these last few years, is it really worth all the pain? And furthermore, many don't like the `six` dependency (https://hugovk.dev/blog/2025/minus-six/).

StanFromIreland commented on tests/resources/data/zoneinfo_v1.json:
I suggest you mark such files as generated in `.gitattributes`.

StanFromIreland commented on tests/resources/update_tzif.py:
```diff
@@ -0,0 +1,241 @@
+    # tzdb-..., which is annoying to deal with. We will delete that and
+    # unpack its contents into `output_dir`
+    for i, subdir in list(enumerate(output_dir.glob("tzdb-*"))):
+        if i > 1:
```
```suggestion
        if i >= 1:
```
Doesn’t this allow for two currently?

StanFromIreland commented on tests/resources/update_tzif.py:
```diff
@@ -0,0 +1,241 @@
+    header_start = 4 + 16
+    header_end = header_start + 24  # 6l == 24 bytes
+    if version < 2:
+        logging.warn(
```
```suggestion
        logging.warning(
```

StanFromIreland commented on tests/resources/update_tzif.py:
```diff
@@ -0,0 +1,241 @@
+            asyncio.create_task(_file_to_v1(tzif_in_file, tzif_v1_file))
+        )
+
+    asyncio.gather(*tasks)
```
Missing `await?`

pganssle committed:
Test with normal tzdata build instead of fat

We were building with fat because dateutil did not support the default
(slim), but now it does, so we can switch to slim. It might make sense
to run the tests with both fat and slim, but I am not convinced that it
would be worth it.

pganssle committed:
fixup! Improve testing for new tzfile implementation

pganssle committed:
fixup! Improve testing for new tzfile implementation
